### PR TITLE
Reused replayer client connections

### PR DESCRIPTION
**Pull request by gregschohn**
### Description
MIGRATIONS-1250. Setup NettyPacketToHttpConsumer to reuse channels...
and to cache them so that they're associated with the original connection ids. That enables us to honor keep-alived connections when doing replays.

To support that, a lot of cleanup and some bugfixes had to happen. The NettyPacketToHttpConsumer initiates the SslHandler and handles the handshake similarly to how the capture proxy does it. Also, like the capture proxy, autoread is now set on the channel once the handlers are in place. finalizeRequest() is responsible for returning a CompletableFuture that will tear down the pipeline as its last (asynchronous) step so that when the channel is passed back into a new NettyPacketToHttpConsumer, the pipeline can be reinitialized (for everything post SslHandler, if applicable).

I've also added a couple helper functions to Utils to support converting a sequence of packets to a string, either by printing them out within brackets ("[]") or by pumping them through an EmbeddedChannel with Http decoders. There's a helper function and a ThreadLocal to manage this more easily for users using toString(), etc that wouldn't normally be able to pass flags into functions.

* Category Enhancement/New feature
* Why these changes are required? To more accurately match the interactions that the source cluster was undergoing.
* What is the old behavior before changes and new behavior after changes?  Every request had a new connection, now requests that shared connection to the source server will share a connection to the target.

### Issues Resolved
[MIGRATIONS-1250](https://opensearch.atlassian.net/browse/MIGRATIONS-1250)

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Unit tests and dockerCompose + OSB

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/RequestPipelineOrchestrator.java [hunk 1]
```diff
@@ -147,6 +147,6 @@ void addBaselineHandlers(ChannelPipeline pipeline) {
     }
 
     private void addLoggingHandler(ChannelPipeline pipeline, String name) {
-        PIPELINE_LOGGING_OPTIONAL.ifPresent(logLevel->pipeline.addLast(new LoggingHandler(name, logLevel)));
+        PIPELINE_LOGGING_OPTIONAL.ifPresent(logLevel->pipeline.addLast(new LoggingHandler("t"+name, logLevel)));
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/RequestPipelineOrchestrator.java:150]: "transform". The other netty logging handlers for network sends are prefaced with an 'n'.  I just wanted a short disambiguator.

File: TrafficCapture/trafficReplayer/src/test/resources/log4j2.properties [hunk 2]
```diff
@@ -1,7 +1,7 @@
 status = error 
 
 # Root logger options
-rootLogger.level = debug
+rootLogger.level = info
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/test/resources/log4j2.properties:4]: No, this was inadvertant that I left that way.  I'm happy to revert if you feel that debug is useful.
Generally, I don't check the logs with gradle builds and intelliJ does a good job for me personally to keep the logs well-partitioned, so I'm fine to set it either way.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/AggregatedRawResponse.java [hunk 3]
```diff
@@ -35,6 +39,7 @@ public static class Builder {
         private final ArrayList<byte[]> requestPackets;
         private final ArrayList<AbstractMap.SimpleEntry<Instant, byte[]>> receiptTimeAndResponsePackets;
         private final Instant requestSendTime;
+        protected Throwable error;
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/AggregatedRawResponse.java:42]: We don't today.  There was a spot within the completable future wirings that made me realize that we could be eating an exception.  At least at the source side, that shouldn't happen.  I can see this coming into importance more once we're focusing on retry policies.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java [hunk 4]
```diff
@@ -138,38 +149,38 @@ public CONNECTION_STATUS addObservationToAccumulation(String nodeId, Accumulatio
         if (observation.hasRead()) {
             rotateAccumulationOnReadIfNecessary(connectionId, accum);
             assert accum.state == Accumulation.State.NOTHING_SENT;
-            log.trace("Adding request data for accum[" + connectionId + "]=" + accum);
+            log.atTrace().setMessage(()->"Adding request data for accum[" + connectionId + "]=" + accum).log();
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java:152]: For this case, placeholders may even be quicker.  However, I'd prefer to canonicalize on this fluent approach for two reasons.  It's easy to be off-by-one with placeholders and the arguments sent for the placeholders are still eagerly evaluated.  This won't be an issue all that often, but if we're doing something that has a function call evaluation, the price could be really high.  The downside of this fluent approach is that we're creating an additional object (though we create a lot of additional objects for lambdas already).

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java [hunk 5]
```diff
@@ -1,44 +1,166 @@
...
+        if (channelsFuture != null) {
+            closeChannel(channelsFuture);
+            connectionId2ChannelCache.invalidate(connId);
+        }
     }
 
-    public void stopGroup() {
-        eventLoopGroup.shutdownGracefully();
+    private DiagnosticTrackableCompletableFuture<String, Channel>
+    closeChannel(ChannelFuture channelsFuture) {
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java:150]: I'll change this to closeClientConnectionChannel

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java [hunk 6]
```diff
@@ -1,44 +1,166 @@
...
+            @Override
+            public ChannelFuture load(final String s) {
+                numConnectionsCreated.incrementAndGet();
+                log.trace("creating connection future");
+                return NettyPacketToHttpConsumer.createClientConnection(eventLoopGroup, sslContext, serverUri, s);
+            }
+        });
     }
 
     NettyPacketToHttpConsumer createNettyHandler(String diagnosticLabel) {
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java:62]: It's constructed in create(), which uses it as an argument.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lewijacn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java:152]: Yeah I think the placeholder is a bit cleaner so I have a slight tendency toward. Even with function calls though couldn't we wrap in a lambda to achieve the same? All that said I think both are better than the tax we have now

**Review by lewijacn — APPROVED** [review]

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/PacketToTransformingHttpHandlerFactory.java:62]: Ahhh - yes, I think that you're correct.  Both are going away for the future work to support metering and backpressure.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java:152]: I think that both format strings and setMessage lambdas are fair to use as long as we understand that < 10% of the time, we'll actually want/need to use one over the other for performance reasons.  In all the other cases, it's likely an artistic judgment to determine which has better maintainability.

**Pull request closed without merging**